File: tests/package-management/python-tests/extension_hive_tests.py
import logging
import os
import shutil
import tempfile
from typing import cast
import uuid
from unittest.mock import Mock

import pytest
from apollo3zehn_package_management import PackageReference
from apollo3zehn_package_management._services import ExtensionHive


@pytest.mark.asyncio
async def can_instantiate_extensions_test():
    
    extension_folder_path = "tests/resources/test-extension"

    # create restore folder
    restore_root = os.path.join(tempfile.gettempdir(), f"PackageManagement.Tests.{uuid.uuid4()}")
    os.makedirs(restore_root)

    try:

        # load packages
        version = "v0.1.0"

        package_reference = PackageReference(
            provider="local",
            configuration={
                "path": extension_folder_path,
                "version": version,
                "entrypoint": "src",
                "import": "my_package.my_module"
            }
        )

        package_reference_map = {
            uuid.uuid4(): package_reference
        }

        hive = ExtensionHive[logging.Logger](restore_root, Mock())
        await hive.load_packages(package_reference_map)

        # instantiate
        loggerType = hive.get_extension_type("my_package.my_module.MyLogger")
        logger =  cast(logging.Logger, loggerType())

        with pytest.raises(Exception, match=r"c\[_\]"):
            assert logger.setLevel(0)

    finally:

        shutil.rmtree(restore_root)

File: tests/package-management/dotnet-tests/ExtensionHiveTests.cs
// MIT License
// Copyright (c) [2024] [Apollo3zehn]

using Apollo3zehn.PackageManagement;
using Apollo3zehn.PackageManagement.Services;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;
using Moq;
using Xunit;

namespace Services;

public class ExtensionHiveTests
{
    [Fact]
    public async Task CanInstantiateExtensions()
    {
        var extensionFolderPath = "../../../../tests/resources/test-extension";

        // create restore folder
        var restoreRoot = Path.Combine(Path.GetTempPath(), $"PackageManagement.Tests.{Guid.NewGuid()}");
        Directory.CreateDirectory(restoreRoot);

        try
        {
            // load packages
            var pathsOptions = Mock.Of<IPackageManagementPathsOptions>();

            Mock.Get(pathsOptions)
                .SetupGet(pathsOptions => pathsOptions.Packages)
                .Returns(restoreRoot);

            var loggerFactory = Mock.Of<ILoggerFactory>();

            Mock.Get(loggerFactory)
                .Setup(loggerFactory => loggerFactory.CreateLogger(It.IsAny<string>()))
                .Returns(NullLogger.Instance);

            var hive = new ExtensionHive<ILogger>(pathsOptions, NullLogger<ExtensionHive<ILogger>>.Instance, loggerFactory);
            var version = "v0.1.0";

            var packageReference = new PackageReference(
                Provider: "local",
                Configuration: new Dictionary<string, string>
                {
                    ["path"] = extensionFolderPath,
                    ["version"] = version,
                    ["entrypoint"] = "test-extension.csproj"
                }
            );

            var packageReferenceMap = new Dictionary<Guid, PackageReference>
            {
                [Guid.NewGuid()] = packageReference
            };

            await hive.LoadPackagesAsync(packageReferenceMap, new Progress<double>(), CancellationToken.None);

            // instantiate
            var type = hive.GetExtensionType("Foo.MyLogger");
            var logger = (ILogger)Activator.CreateInstance(type)!;

            Assert.NotNull(logger);
            Assert.False(logger.IsEnabled(LogLevel.Trace));
            Assert.True(logger.IsEnabled(LogLevel.Debug));
        }
        finally
        {
            try
            {
                Directory.Delete(restoreRoot, recursive: true);
            }
            catch { }
        }
    }
}